### PR TITLE
TSV-500 Refactor auditing and conditions.  Replace sticky routing implementation, remove hourly limits.

### PR DESCRIPTION
**Pull request by dclifton-hmrc**

### CODE REVIEW

**Comment by cellis-ee** [comment]
please don't merge the above before TSV-497, and before @obura says it's okay

File: app/model/Conditions.scala [hunk 1]
```diff
@@ -17,124 +17,68 @@
...
+      Future.successful(!rc.request_.session.data.contains("token") && rc.credId.isEmpty)
 
-  override val auditType = Some(AFFINITY_GROUP_AVAILABLE)
-}
+    val hasAnyBusinessEnrolmentF: ConditionPredicate = rc =>
+      rc.activeEnrolmentKeys.map(_.intersect(rc.businessEnrolments).nonEmpty)
 
-trait HasAnyBusinessEnrolment extends Condition {
-  def businessEnrolments: Set[String]
+    val SAReturnAvailableF: ConditionPredicate = rc =>
```
Comment (cellis-ee) [inline app/model/Conditions.scala]: `SA` should be lowercase at the start of this variable name

File: app/controllers/RouterController.scala [hunk 2]
```diff
@@ -57,33 +57,41 @@ trait RouterController extends FrontendController with Actions {
...
+    def sendAuditEvent(auditInfo: AuditInfo, throttledLocation: Location)(implicit authContext: AuthContext, request: Request[AnyContent], hc: HeaderCarrier) = {
+      val auditEvent = auditInfo.toAuditEvent(throttledLocation)
       auditConnector.sendEvent(auditEvent)
-      Logger.debug(s"Routing decision summary: ${auditEvent.detail \ "reasons"}")
+      val reasons: JsValue = (auditEvent.detail \ "reasons").getOrElse(JsNull)
+      Logger.debug(s"Routing decision summary: ${Json.stringify(reasons)}")
+    }
+
+    val destinationAfterRulesApplied = ruleEngine.getLocation(ruleContext)
+    val finalResult = throttlingService.throttle(destinationAfterRulesApplied, ruleContext).run
```
Comment (cellis-ee) [inline app/controllers/RouterController.scala]: this isn't the final result, because we do more processing on it afterwards

File: app/controllers/RouterController.scala [hunk 3]
```diff
@@ -57,33 +57,41 @@ trait RouterController extends FrontendController with Actions {
...
+    val futureAuditInfo = finalResult map { case (auditInfo, _) => auditInfo }
+    val futureFinalDestination = finalResult map { case (_, finalDestination) => finalDestination }
+
+    val destination = for {
+      finalDestination <- futureFinalDestination
+      auditInfo <- futureAuditInfo
+    } yield {
+      sendAuditEvent(auditInfo, finalDestination)
+      metricsMonitoringService.sendMonitoringEvents(auditInfo, finalDestination)
+      analyticsEventSender.sendEvents(finalDestination.name, auditInfo)
```
Comment (cellis-ee) [inline app/controllers/RouterController.scala]: the two lines above this take `auditInfo` and `finalDestination` in the opposite order. can we standardise this one way or the other?

File: app/controllers/internal/AccountTypeController.scala [hunk 4]
```diff
@@ -66,20 +63,17 @@ trait AccountTypeController extends FrontendController with Actions {
...
 
-  def createAuditContext(): TAuditContext
+  def createRuleContext(credId: String)(implicit request: Request[AnyContent], hc: HeaderCarrier): RuleContext
 
-  def createRuleContext(credId: String)(implicit hc: HeaderCarrier): RuleContext
-
-  def accountTypeForCredId(credId: String): Action[AnyContent] = Action.async { implicit request =>
-    val ruleContext = createRuleContext(credId)
+  def accountTypeForCredId(credId: String) = Action.async { implicit request =>
+    val ruleContext: RuleContext = createRuleContext(credId)
```
Comment (cellis-ee) [inline app/controllers/internal/AccountTypeController.scala]: unnecessary type

File: app/engine/AuditInfo.scala [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+  lazy val transactionNames = Map(
+    Locations.PersonalTaxAccount -> "sent to personal tax account",
+    Locations.BusinessTaxAccount -> "sent to business tax account"
+  )
+
+  def toAuditEvent(location: Location)(implicit hc: HeaderCarrier, authContext: AuthContext, request: Request[AnyContent]): ExtendedDataEvent = {
+    this match {
+      case AuditInfo(routingReasons, ruleApplied, throttlingInfo) =>
+        val accounts: Accounts = authContext.principal.accounts
+        val accountMap = accounts.toMap
```
Comment (cellis-ee) [inline app/engine/AuditInfo.scala]: unnecessary line - add `.toMap` to end of previous line

File: app/engine/AuditInfo.scala [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+sealed trait TAuditInfo {
+  lazy val transactionNames = Map(
+    Locations.PersonalTaxAccount -> "sent to personal tax account",
+    Locations.BusinessTaxAccount -> "sent to business tax account"
+  )
+
+  def toAuditEvent(location: Location)(implicit hc: HeaderCarrier, authContext: AuthContext, request: Request[AnyContent]): ExtendedDataEvent = {
+    this match {
+      case AuditInfo(routingReasons, ruleApplied, throttlingInfo) =>
+        val accounts: Accounts = authContext.principal.accounts
```
Comment (cellis-ee) [inline app/engine/AuditInfo.scala]: unnecessary type

File: app/engine/condition.scala [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+
+case class Pure[C](predicate: C => Future[Boolean], routingReason: RoutingReason) extends Condition[C] with WithRoutingReason
+case class And[C](c1: Condition[C], c2: Condition[C]) extends Condition[C]
+case class Or[C](c1: Condition[C], c2: Condition[C]) extends Condition[C]
+case class Not[C](condition: Condition[C]) extends Condition[C]
+case class When[C](condition: Condition[C])
+
+sealed trait Condition[C] extends Expr[C, Boolean] {
+
+  def evaluate[A <: AuditInfo: Semigroup](context: C): ConditionResult = {
```
Comment (cellis-ee) [inline app/engine/condition.scala]: is this behaviour well-tested?

File: app/engine/package.scala [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+      case Left(aa) => tailRecM(aa)(f)
+    }
+
+    override def map[A, B](fa: Future[A])(f: (A) => B): Future[B] = fa map f
+  }
+
+  type ConditionResult = WriterT[Future, AuditInfo, Boolean]
+  type RuleResult = WriterT[Future, AuditInfo, Option[Location]]
+  type EngineResult = WriterT[Future, AuditInfo, Location]
+  val emptyRuleResult: RuleResult = WriterT(Future.successful((AuditInfo.Empty, None)))
```
Comment (cellis-ee) [inline app/engine/package.scala]: please add a line to separate the `val` from the `type`s

File: app/model/Location.scala [hunk 9]
```diff
@@ -16,57 +16,39 @@
...
 
-  def twoStepVerificationRequired(queryString: Map[String, String]) = locationFromConf("two-step-verification-required").copy(queryParams = queryString)
+    def getString(key: String): String = {
+      appConfig.getLocationConfig(locationName, key)
+        .getOrElse(throw new RuntimeException(s"key '$key' not configured for location '$locationName'"))
+    }
 
-  lazy val all = List(PersonalTaxAccount, BusinessTaxAccount)
+    val name = getString("name")
+    val url = getString("url")
```
Comment (cellis-ee) [inline app/model/Location.scala]: vals can be inlined?

File: app/services/MetricsMonitoringService.scala [hunk 10]
```diff
@@ -34,22 +35,24 @@ trait MetricsMonitoringService {
...
-  def sendMonitoringEvents(auditContext: TAuditContext, throttledLocation: Location)(implicit request: Request[AnyContent], hc: HeaderCarrier): Future[Unit] = {
+  def sendMonitoringEvents(auditInfo: AuditInfo, throttledLocation: Location)(implicit request: Request[AnyContent], hc: HeaderCarrier): Future[Unit] = {
 
     Future {
-      val destinationNameBeforeThrottling = auditContext.getThrottlingDetails.get("destination-name-before-throttling")
-      val destinationNameAfterThrottling = throttledLocation.name
-      val throttleKey = if (destinationNameBeforeThrottling.isDefined && destinationNameBeforeThrottling.get != destinationNameAfterThrottling) {
+      val destinationNameBeforeThrottling: Option[String] = auditInfo.throttlingInfo.map(_.initialDestination.name)
+      val destinationNameAfterThrottling: String = throttledLocation.name
+      val throttleKey: String = if (destinationNameBeforeThrottling.isDefined && destinationNameBeforeThrottling.get != destinationNameAfterThrottling) {
```
Comment (cellis-ee) [inline app/services/MetricsMonitoringService.scala]: unnecessary types?

File: it/router/RouterAuditFeature.scala [hunk 11]
```diff
@@ -501,18 +591,85 @@ class RouterAuditFeature extends StubbedFeatureSpec with CommonStubs {
...
+
+    And("user is sent to BTA an audit event should be sent")
+    verify(postRequestedFor(urlMatching("^/write/audit.*$")))
+
+    And("the audit event raised should be the expected one")
+    val expectedReasons = toJson(emptyRoutingReasons + (IS_A_VERIFY_USER.key -> "true"))
+    val expectedTransactionName = "sent to business tax account"
+    val expectedThrottlingDetails = ThrottlingDetails(
+      enabled = true,
+      percentage = "100",
```
Comment (cellis-ee) [inline it/router/RouterAuditFeature.scala]: this is the only test i can see where `percentage` is anything other than `"-"`. Should there be a test where `percentage` is eg. 50, and then tests where `ThrottlingDetails` contains both `true` and `false` for `throttled` ?

File: it/router/RouterSaUnresponsiveFeature.scala [hunk 12]
```diff
@@ -43,17 +42,21 @@ class RouterSaUnresponsiveFeature extends StubbedFeatureSpec with CommonStubs {
       Then("the user should be routed to BTA Home Page")
       on(BtaHomePage)
 
-      And("the authority object should be fetched once for AuthenticatedBy")
-      verifyAuthorityObjectIsFetched()
+      eventually {
```
Comment (cellis-ee) [inline it/router/RouterSaUnresponsiveFeature.scala]: i would have thought that fetching the auth object, and the following calls inside this `eventually` block occur before the user eventually arrives at the `BtaHomePage`. If so, that suggests that the `eventually {` line should be moved upwards by a few lines

File: it/router/RouterSaUnresponsiveFeature.scala [hunk 13]
```diff
@@ -43,17 +42,21 @@ class RouterSaUnresponsiveFeature extends StubbedFeatureSpec with CommonStubs {
...
-      verify(0, getRequestedFor(urlEqualTo("/user-details-uri")))
+        And("user's enrolments should be fetched from Auth")
+        verify(getRequestedFor(urlEqualTo("/auth/enrolments-uri")))
 
-      And("sa returns should be fetched from Sa micro service")
-      verify(getRequestedFor(urlEqualTo(s"/sa/individual/$saUtr/return/last")))
+        And("user's details should be fetched from User Details")
+        verify(0, getRequestedFor(urlEqualTo("/user-details-uri")))
+
+        And("sa returns should be fetched from Sa micro service")
```
Comment (cellis-ee) [inline it/router/RouterSaUnresponsiveFeature.scala]: casing of `SA`

File: it/router/RouterSaUnresponsiveFeature.scala [hunk 14]
```diff
@@ -74,17 +77,21 @@ class RouterSaUnresponsiveFeature extends StubbedFeatureSpec with CommonStubs {
...
+        verify(getRequestedFor(urlEqualTo("/auth/enrolments-uri")))
 
-      And("user's enrolments should be fetched from Auth")
-      verify(getRequestedFor(urlEqualTo("/auth/enrolments-uri")))
+        And("user's details should be fetched from User Details")
+        verify(0, getRequestedFor(urlEqualTo("/user-details-uri")))
 
-      And("user's details should be fetched from User Details")
-      verify(0, getRequestedFor(urlEqualTo("/user-details-uri")))
+        And("Sa micro service should not be invoked")
```
Comment (cellis-ee) [inline it/router/RouterSaUnresponsiveFeature.scala]: casing of `SA`

File: it/router/RouterSaUnresponsiveFeature.scala [hunk 15]
```diff
@@ -74,17 +77,21 @@ class RouterSaUnresponsiveFeature extends StubbedFeatureSpec with CommonStubs {
       Then("the user should be routed to BTA Home Page")
       on(BtaHomePage)
 
-      And("the authority object should be fetched once for AuthenticatedBy")
-      verifyAuthorityObjectIsFetched()
+      eventually {
```
Comment (cellis-ee) [inline it/router/RouterSaUnresponsiveFeature.scala]: as comment on `eventually` above

File: test/controllers/RouterControllerSpec.scala [hunk 16]
```diff
@@ -16,218 +16,126 @@
...
-
-      //and
-      when(mockThrottlingService.throttle(eqTo(expectedLocation), eqTo(auditContext), eqTo(ruleContext))(eqTo(fakeRequest), any[ExecutionContext])).thenReturn(expectedLocation)
-
-      val controller = new TestRouterController(ruleEngine = ruleEngineStubReturningNoneLocation, throttlingService = mockThrottlingService)
+  "router controller" should {
 
-      //when
-      val route = controller.route
+    "return location" in new Setup {
```
Comment (cellis-ee) [inline test/controllers/RouterControllerSpec.scala]: the old test file had a few more tests (default location, missing origin) - might be worth checking that none of these are useful

File: test/engine/ConditionsSpec.scala [hunk 17]
```diff
@@ -55,27 +55,20 @@ class ConditionsSpec extends UnitSpec with MockitoSugar with WithFakeApplication
 
     forAll(scenarios) { (scenario: String, enrolments: Set[String], expectedResult: Boolean) =>
       s"be true whether the user has any business enrolments - scenario: $scenario" in {
```
Comment (cellis-ee) [inline test/engine/ConditionsSpec.scala]: i know this line hasn't changed in this work, but usually the word 'whether' comes with an alternative - eg. 'whether or not'

File: test/engine/RuleEngineSpec.scala [hunk 18]
```diff
@@ -16,124 +16,90 @@
...
-
-  val trueLocation: Location = evaluateUsingPlay(Location("/true", "true"))
-  val trueRule = When(BooleanCondition(true)).thenGoTo(trueLocation) withName "true-rule"
-  val falseLocation: Location = evaluateUsingPlay(Location("/false", "false"))
-  val falseRule = When(BooleanCondition(false)).thenGoTo(falseLocation) withName "false-rule"
-
-  val theDefaultLocation = Location("default-location", "/default-location")
+  "rule engine" should {
 
+    "rule matched" in new Setup {
```
Comment (cellis-ee) [inline test/engine/RuleEngineSpec.scala]: The sentence "rule engine should rule matched" is not good English.

File: test/engine/RuleEngineSpec.scala [hunk 19]
```diff
@@ -16,124 +16,90 @@
...
-
-      auditContext.ruleApplied shouldBe trueRule.name
+      evaluatedReasons shouldBe Map(Reason("alwaysTrue") -> Some(true))
+      location shouldBe location1
     }
 
-    "evaluate empty rules and return with default location" in {
-      implicit lazy val request = FakeRequest()
-      implicit lazy val hc: HeaderCarrier = HeaderCarrier.fromHeadersAndSession(request.headers)
+    "rule not matched" in new Setup {
```
Comment (cellis-ee) [inline test/engine/RuleEngineSpec.scala]: as above

File: test/model/LocationsSpec.scala [hunk 20]
```diff
@@ -16,87 +16,58 @@
...
       }
-    }
 
-    "throw RuntimeException when location url is missing from configuration" in new Setup {
+      val caught = intercept[RuntimeException] {
+        locations.TaxAccountRouterHome
+      }
 
-      implicit val fakeApplication = FakeApplication(additionalConfiguration = additionalConfiguration - "locations.some-location.url")
+      caught shouldNot be(null)
```
Comment (cellis-ee) [inline test/model/LocationsSpec.scala]: should you make any further assertions on the caught exception? to make sure that it's the one you're expecting

File: test/services/LocationConfigurationFactorySpec.scala [hunk 21]
```diff
@@ -0,0 +1,77 @@
...
+      when(mockAppConfig.getThrottlingConfig("name1")).thenReturn(testThrottlingConfig)
+
+      val location = Location("name1", "some-url")
+
+      val result = locationConfigurationFactory.configurationForLocation(location, fakeRequest)
+
+      result shouldBe testThrottlingConfig
+    }
+
+    "retrieve appropriate configuration for PTA" in new Setup {
```
Comment (cellis-ee) [inline test/services/LocationConfigurationFactorySpec.scala]: an addition to the test name with something like "without token in session" may be helpful. I had to look up why `-verify` was added to the string below

File: test/services/MetricsMonitoringServiceSpec.scala [hunk 22]
```diff
@@ -17,49 +17,57 @@
...
-        val mockAuditContext = mock[TAuditContext]
+        import engine.RoutingReason._
 
-        val mockThrottledLocation = mock[Location]
-        when(mockThrottledLocation.name).thenReturn(destinationNameAfterThrottling)
+        // given
+        val c1 = IS_A_VERIFY_USER
+        val c2 = IS_A_GOVERNMENT_GATEWAY_USER
+        val c3 = IS_IN_A_PARTNERSHIP
+        val c4 = IS_SELF_EMPLOYED
```
Comment (cellis-ee) [inline test/services/MetricsMonitoringServiceSpec.scala]: can these conditions be named something more expressive? and also can they be moved to `Setup` (along with the 'meter' mocking below) ?

File: app/engine/condition.scala (hunk 7 above)
Comment (dclifton-hmrc) [inline app/engine/condition.scala]: no, the expected behaviour of the AND, OR and NOT conditions are not well defined.  There is a test hole.

File: test/controllers/RouterControllerSpec.scala (hunk 16 above)
Comment (dclifton-hmrc) [inline test/controllers/RouterControllerSpec.scala]: craig says "that's ok then".

File: test/model/LocationsSpec.scala (hunk 20 above)
Comment (dclifton-hmrc) [inline test/model/LocationsSpec.scala]: done

File: app/engine/condition.scala (hunk 7 above)
Comment (dclifton-hmrc) [inline app/engine/condition.scala]: done, see a34f640d6416d5854da32723770e9590e755b41e
